heartbeat/hash_detector.go

```diff
@@ -0,0 +1,102 @@
+// NewHashDetector as seen below. It is created here not only to simplify the
+// method signature of that function, but also to provide additional
+// documentation about the nature of each parameter.
+type HashDetectorParam struct {
+	// Pool is a *redis.Pool capable of making new connections to a
+	// particular Redis instance (or cluster). It is used in the `func
+	// Detect` to initialize a connection to Redis.
+	Pool *redis.Pool
+
+	// MaxAge is the maximum amount of time that can pass between the last
+	// update of a particular item in the Hash before it is considered
+	// "dead".
+	MaxAge time.Duration
+
+	// Location is the keyspace in which to search for items in the Hash. It
+	// need not be initialized with a name, only the keyspace in which to
+	// search.
+	Location *keys.Location
+}
+
+// HashDetector is an implementation of the Detector interface that searches for
+// items in a Redis HASH datastructure and returns them as "expired" or "dead"
+// based on whether or not a given time interval has passed between their last
+// update, and now.
+type HashDetector struct {
+	// MaxAge is the maximum amount of time that can pass between now and
+	// the last update of a particular item in the HASH for it to still be
+	// considered "alive".
+	MaxAge time.Duration
+
+	// pool is a *redis.Pool, responsible for making connections to the
+	// Redis instance in which to check for items.
+	pool *redis.Pool
+
+	// location is the location in the Redis keyspace in which to search.
+	location *keys.Location
+}
+
+// NewHashDetector initializes and returns a new HashDetector with the given
+// parameters as specified and documented in the HashDetectorParam.
+func NewHashDetector(param HashDetectorParam) HashDetector {
+	return SimpleDetector{
+		MaxAge:   param.MaxAge,
+		pool:     param.Pool,
+		location: param.Location,
+	}
+}
+
+// Pool implements the `func Pool` defined on the `type Detector interface`. It
+// returns a reference to the Pool used to make connections to Redis.
+func (d *SimpleDetector) Pool() *redis.Pool {
+	return d.pool
+}
+
+// Location implements the `func Location` defined on the `type Detector
+// interface`. It returns the location in which to search for items.
+func (d *SimpleDetector) Location() *keys.Location {
+	return d.location
+}
+
+// Detect implements the `func Detect` defined on the `type Detector
+// interface`. It scans iteratively over the given keyspace and looks for items
+// that have expired. It returrns the IDs of those items in the `expired` return
+// param, and halts (as well as returns) any errors that it encounters while
+// executing.
+func (d *SimpleDetector) Detect() (expired []string, err error) {
+	now := time.Now()
+	redis := d.Pool().Get()
+	defer redis.Close()
+
+	reply, err := redis.StringMap(redis.Do("HGETALL",
+		d.Location().Keyspace()))
+
+	if err != nil {
+		return
+	}
+
+	// TODO(ttaylorr): split this into the worker-pattern and have
+	// individual goroutines process it.
```

Comment:
unless you have upwards of 10k heartbeats happening, this should not be an issue...